proc/groupby.go

```diff
@@ -188,72 +202,74 @@ func typeKey(rowkeys []zeek.TypedEncoding) string {
 	return b.String()
 }
 
-func (g *GroupByAggregator) createRow(ts nano.Ts, rowkeys []zeek.TypedEncoding) *GroupByRow {
+func (g *GroupByAggregator) createRow(id int, ts nano.Ts, vals zval.Encoding) *GroupByRow {
 	// Make a deep copy so the caller can reuse the underlying arrays.
-	v := make([]zeek.TypedEncoding, len(rowkeys))
-	copy(v, rowkeys)
-	key := typeKey(v)
-	_, ok := g.typeCols[key]
-	if !ok {
-		g.typeCols[key] = v
-	}
+	v := make(zval.Encoding, len(vals))
+	copy(v, vals)
 	return &GroupByRow{
-		rowKeys: v,
-		ts:      ts,
-		columns: compile.Row{Defs: g.reducerDefs},
+		id:       id,
+		keyvals:  v,
+		ts:       ts,
+		reducers: compile.Row{Defs: g.reducerDefs},
 	}
 }
 
 // XXX this could be made more efficient by using exporting zval.Encoding.build
 // and using it rowkey.Body.Build() and do something smarter with the zeek type
 // strings... otherwise we are sending lots of strings to the GC on each record
 // defeating the purpose of g.cacheKey.
-func (g *GroupByAggregator) key(key []byte, rowkeys []zeek.TypedEncoding) ([]byte, error) {
-	if len(rowkeys) > 0 {
-		key = append(key, rowkeys[0].Type.String()...)
-		for _, rowkey := range rowkeys[1:] {
-			key = append(key, ':')
-			key = append(key, rowkey.Type.String()...)
-		}
-		for _, rowkey := range rowkeys {
-			key = append(key, ':')
-			key = append(key, rowkey.String()...)
-		}
-	}
-	return key, nil
+func encodeInt(dst zval.Encoding, v int) {
+	dst[0] = byte(v >> 24)
+	dst[1] = byte(v >> 16)
+	dst[2] = byte(v >> 8)
+	dst[3] = byte(v)
 }
 
 // Consume takes a record and adds it to the aggregation. Records
 // successively passed to Consume are expected to have timestamps in
 // monotonically increasing or decreasing order determined by g.reverse.
 func (g *GroupByAggregator) Consume(r *zson.Record) error {
-	// Extract the list of groupby expressions.  Re-use the array
-	// stored in cacheRowKeys to avoid re-allocating on every record.
-	var rowkeys []zeek.TypedEncoding
-	if g.cacheRowKeys != nil {
-		rowkeys = g.cacheRowKeys[:0]
-	}
-	for _, key := range g.keys {
-		rowkey := key.resolver(r)
-		if rowkey.Type != nil {
-			rowkeys = append(rowkeys, rowkey)
-		}
+	// First check if we've seen this descriptor before and if not
+	// build an entry for it.
+	id := r.Descriptor.ID
+	keycols, ok := g.keyColsTable[id]
+	if !ok {
+		keycols = newKeyCols(r.Descriptor, g.keys)
+		g.keyColsTable[id] = keycols
+	} else if keycols == nil {
```

Comment:
This shouldn't be else right?  Otherwise, the first record of a type that doesn't include all the keys still gets past this block...